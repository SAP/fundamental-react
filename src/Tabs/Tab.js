import classnamesBind from 'classnames/bind';
import Icon from '../Icon/Icon';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import iconStyles from 'fundamental-styles/dist/icon.css';
import tabStyles from 'fundamental-styles/dist/tabs.css';

const classnames = classnamesBind.bind({
    ...iconStyles,
    ...tabStyles
});

const Tab = React.forwardRef(({ title, glyph, id, selected, onClick,
    tabContentProps, linkProps, index, className, cssNamespace, ...props }, ref) => {

    const tabClasses = classnames(
        className,
        `${cssNamespace}-tabs__item`
    );

    return (
        <li
            {...props}
            aria-controls={id}
            aria-selected={selected}
            className={tabClasses}
            key={id}
            ref={ref}
            role='tab'>
            <a
                {...linkProps}
                className={classnames(`${cssNamespace}-tabs__link`)}
                href={`#${id}`}
                onClick={(event) => onClick(event, index)}>
                {glyph ?
                    <span className={classnames('fd-tabs__icon')}>
                        <Icon ariaLabel={title} glyph={glyph} />
                    </span>
                    :
                    <span className={classnames(`${cssNamespace}-tabs__tag`)}>{title}</span>
                }
            </a>
        </li>
    );
});

Tab.displayName = 'Tab';

Tab.defaultProps = {
    onClick: () => { }
};

Tab.propTypes = {
    /** Localized text for the heading */
    title: PropTypes.string.isRequired,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** The icon to include. See the icon page for the list of icons */
    glyph: PropTypes.string,
    /** Value for the `id` attribute on the element */
    id: PropTypes.string,
    /** Internal use only */
    index: PropTypes.number,
    /** Additional props to be spread to the tab\'s `<a>` element */
    linkProps: PropTypes.object,
    /** Internal use only */
    selected: PropTypes.bool,
    /** Additional props to be spread to the tab content\'s `<div>` element */
    tabContentProps: PropTypes.object,
    /** Internal use only */
    onClick: PropTypes.func
};

export default withStyles(Tab);
