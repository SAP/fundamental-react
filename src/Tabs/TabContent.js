import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

export const TabContent = (props) => {
    const { children, id, selected, className, ...tabContentProps } = props;

    // css classes for tab panels
    const tabPanelClasses = classnames(
        'fd-tabs__panel',
        className
    );

    return (
        <div
            {...tabContentProps}
            aria-expanded={selected}
            className={tabPanelClasses}
            id={id}
            role='tabpanel' >
            {children}
        </div >
    );
};
TabContent.displayName = 'TabContent';

TabContent.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    id: PropTypes.string,
    selected: PropTypes.bool
};

TabContent.propDescriptions = {
    children: 'Content to render when tab is selected.'
};
