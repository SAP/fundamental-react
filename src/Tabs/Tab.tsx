import * as React from 'react';
import { ReactNode } from 'react';
import { TabKey } from "./Tabs";
import { StatelessComponent } from 'enzyme';

export interface TabProps {
    /**
     * A unique (per Tabs component) identifier for the tab.
     */
    key: TabKey;

    /**
     * The content to render in the tab title
     */
    title: ReactNode;

    /**
     * If the tab is disabled. If disabled, it cannot be activated.
     */
    disabled?: boolean;

    /**
     * The tab content
     */
    children?: ReactNode;
}

/** 
 * A Tab to render within the Tabs component.
 */
export const Tab: StatelessComponent<TabProps> = (props: TabProps) => (
    <React.Fragment>
        {props.children || null}
    </React.Fragment>
);