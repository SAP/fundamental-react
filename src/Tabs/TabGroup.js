import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import { TAB_SIZES } from '../utils/constants';
import TabContent from './_TabContent';
import withStyles from '../utils/withStyles';
import React, { Component } from 'react';
import styles from 'fundamental-styles/dist/tabs.css';

const classnames = classnamesBind.bind(styles);

/** A **TabGroup** is a collection of **Tab** components.  Each **Tab** is based on a folder
metaphor and is used to separate content into different sections.
They should be ordered to create a visual hierarchy based on priority. */
class TabGroup extends Component {
    constructor(props) {
        super(props);
        this.state = {
            selectedIndex: props.selectedIndex
        };
    }

    static getDerivedStateFromProps(props, state) {
        const prevProps = state.prevProps || {};
        // Compare the incoming prop to previous prop
        const selectedIndex =
            prevProps.selectedIndex !== props.selectedIndex
                ? props.selectedIndex
                : state.selectedIndex;
        return {
            // Store the previous props in state
            prevProps: props,
            selectedIndex
        };
    }

    // set selected tab
    handleTabSelection = (event, index) => {
        event.preventDefault();
        this.setState({
            selectedIndex: index
        });

        this.props.onTabClick(event, index);
    };

    // clone Tab element
    cloneElement = (child, index) => {
        return (React.cloneElement(child, {
            onClick: this.handleTabSelection,
            selected: this.state.selectedIndex === index,
            index: index
        }));
    }

    // create tab list
    renderTabs = () => {
        return React.Children.toArray(this.props.children).map((child, index) => {
            return this.cloneElement(child, index);
        });
    };

    // create content to show below tab list
    renderContent = () => {
        return React.Children.toArray(this.props.children).map((child, index) => {
            return (
                <TabContent
                    {...child.props.tabContentProps}
                    key={index}
                    selected={this.state.selectedIndex === index}>
                    {child.props.children}
                </TabContent>);
        });
    };

    render() {
        const {
            children,
            className,
            cssNamespace,
            selectedIndex,
            size,
            tabGroupProps,
            onTabClick,
            ...rest } = this.props;

        // css classes to use for tab group
        const tabGroupClasses = classnames(
            `${cssNamespace}-tabs`,
            { [`${cssNamespace}-tabs--${size}`]: size },
            className
        );
        return (
            <React.Fragment>
                <ul {...rest} className={tabGroupClasses}
                    role='tablist'>
                    {this.renderTabs(children)}
                </ul>
                {this.renderContent(children)}
            </React.Fragment>
        );
    }
}

TabGroup.displayName = 'TabGroup';

TabGroup.defaultProps = {
    selectedIndex: 0,
    size: 'l',
    onTabClick: () => { }
};

TabGroup.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** The index of the selected tab */
    selectedIndex: PropTypes.number,
    /** Size of the component: 's',
    'm',
    'l',
    'xl',
    'xxl' */
    size: PropTypes.oneOf(TAB_SIZES),
    tabGroupProps: PropTypes.object,
    /**
     * Callback function; triggered when the user selects a tab by clicking on pressing `enter`.
     *
     * @param {SyntheticEvent} event - React's original SyntheticEvent. See https://reactjs.org/docs/events.html.
     * @param {number} index - selected tab index.
     * @returns {void}
     * */
    onTabClick: PropTypes.func
};

export default withStyles(TabGroup);
