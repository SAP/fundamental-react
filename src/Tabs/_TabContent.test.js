import { mount } from 'enzyme';
import React from 'react';
import Tab from './Tab';
import TabContent from './_TabContent';
import TabGroup from './TabGroup';

describe('<Tabs />', () => {
    const shownTabContent = (
        <TabContent id='4'>
            Lorem ipsum dolor sit amet consectetur adipisicing elit.Dolore et ducimus veritatis officiis amet ? Vitae officia optio dolor exercitationem incidunt magnam non, suscipit, illo quisquam numquam fugiat ? Debitis, delectus sequi ?
        </TabContent>);

    const hiddenTabContent = (
        <TabContent id='5'>
            Lorem ipsum dolor sit amet consectetur adipisicing elit.Dolore et ducimus veritatis officiis amet ? Vitae officia optio dolor exercitationem incidunt magnam non, suscipit, illo quisquam numquam fugiat ? Debitis, delectus sequi ?
        </TabContent>);

    test('check if tab content is shown', () => {
        const wrapper = mount(shownTabContent);
        wrapper.setProps({ 'selected': true });

        const isExpanded = wrapper.getDOMNode().getAttribute('aria-expanded');
        expect(isExpanded).toEqual('true');
    });

    test('check if tab content is hidden', () => {
        const wrapper = mount(hiddenTabContent);
        wrapper.setProps({ 'selected': false });

        const isExpanded = wrapper.getDOMNode().getAttribute('aria-expanded');
        expect(isExpanded).toEqual('false');
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Tab component\'s content component', () => {
            const element = mount(<TabGroup selectedIndex={1}>
                <Tab id='1' tabContentProps={{ 'data-sample': 'Sample' }}
                    title='Tab 1' />
            </TabGroup>
            );

            expect(
                element.find('div.fd-tabs__panel').at(0).getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });
});
