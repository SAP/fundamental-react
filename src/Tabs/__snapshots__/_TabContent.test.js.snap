// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> create TabContent component 1`] = `
<div
  className="fd-tabs__panel"
  id="4"
  role="tabpanel"
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.Dolore et ducimus veritatis officiis amet ? Vitae officia optio dolor exercitationem incidunt magnam non, suscipit, illo quisquam numquam fugiat ? Debitis, delectus sequi ?
</div>
`;

exports[`<Tabs /> create TabContent component 2`] = `
<div
  className="fd-tabs__panel"
  id="5"
  role="tabpanel"
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.Dolore et ducimus veritatis officiis amet ? Vitae officia optio dolor exercitationem incidunt magnam non, suscipit, illo quisquam numquam fugiat ? Debitis, delectus sequi ?
</div>
`;
